// necessary/lib/utilities/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _array = require("../utilities/array");
var _defaults = require("../defaults");
var _constants = require("../constants");
var _fileSystem = require("../utilities/fileSystem");
function _instanceof(left, right) {
    if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
        return right[Symbol.hasInstance](left);
    } else {
        return left instanceof right;
    }
}
var _typeof = function(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
};
if (!globalThis.rc) {
    globalThis.rc = _rc;
}
var rc = globalThis.rc;
var _default = {
    rc: rc
};
exports.default = _default;
var baseExtension = _defaults.DEFAULT_RC_BASE_EXTENSION;
function _rc(environmentName) {
    environmentName = environmentName || environmentNameFromArgv() || environmentNameFromEnv(); ///
    var json = readRCFile();
    var environment = findEnvironment(json, environmentName);
    json = environment; ///
    replaceEnvironmentVariables(json);
    Object.assign(rc, environment);
}
function readRCFile() {
    var rcFilePath = rcFilePathFromNothing(), rcFileContent = (0, _fileSystem).readFile(rcFilePath), json = JSON.parse(rcFileContent);
    return json;
}
function writeRCFile(json) {
    var rcFilePath = rcFilePathFromNothing(), rdFileContent = JSON.stringify(json, null, "  ");
    (0, _fileSystem).writeFile(rcFilePath, rdFileContent);
}
function updateRCFile(addedProperties) {
    for(var _len = arguments.length, deletedPropertyNames = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++){
        deletedPropertyNames[_key - 1] = arguments[_key];
    }
    var json = readRCFile();
    if (addedProperties) {
        Object.assign(json, addedProperties);
    }
    deletedPropertyNames.forEach(function(deletedPropertyName) {
        delete json[deletedPropertyName];
    });
    writeRCFile(json);
}
function checkRCFileExists() {
    var rcFilePath = rcFilePathFromNothing(), rcFileExists = (0, _fileSystem).checkFileExists(rcFilePath);
    return rcFileExists;
}
function createVacuousRCFile() {
    var json = {
        "environments": [
            {}
        ]
    };
    writeRCFile(json);
}
function setRCBaseExtension(rcBaseExtension) {
    baseExtension = rcBaseExtension;
}
Object.assign(_rc, {
    readRCFile: readRCFile,
    writeRCFile: writeRCFile,
    updateRCFile: updateRCFile,
    checkRCFileExists: checkRCFileExists,
    createVacuousRCFile: createVacuousRCFile,
    setRCBaseExtension: setRCBaseExtension
});
function findEnvironment(json, environmentName) {
    var environment1;
    var environments = json.environments;
    if (environmentName === null) {
        var firstEnvironment = (0, _array).first(environments);
        environment1 = firstEnvironment; ///
    } else {
        environment1 = environments.find(function(environment) {
            var name = environment.name;
            if (name === environmentName) {
                return true;
            }
        });
    }
    delete environment1.name;
    return environment1;
}
function rcFilePathFromNothing() {
    var rcFilePath = "./.".concat(baseExtension, "rc");
    return rcFilePath;
}
function environmentNameFromEnv() {
    var environmentName = process.env[_constants.ENVIRONMENT] || null;
    return environmentName;
}
function environmentNameFromArgv() {
    var environmentName = null;
    process.argv.find(function(argument) {
        var matches = argument.match(/--environment=(.+)/), found = matches !== null;
        if (found) {
            var secondMatch = (0, _array).second(matches);
            environmentName = secondMatch;
        }
        return found;
    });
    return environmentName;
}
function replaceEnvironmentVariable(string) {
    var value = null;
    var stringUpperCase = isStringUpperCase(string);
    if (stringUpperCase) {
        var name = string; ///
        value = process.env[name] || null;
    }
    return value;
}
function replaceEnvironmentVariables(environment) {
    var json = environment;
    var jsonArray = isJSONArray(json), jsonObject = isJSONObject(json);
    if (false) {
    ///
    } else if (jsonArray) {
        var array = json, length = array.length;
        for(var index = 0; index < length; index++){
            var json1 = array[index], jsonString = isJSONString(json1);
            if (jsonString) {
                var string = json1, value = replaceEnvironmentVariable(string);
                if (value !== null) {
                    array[index] = value;
                }
            }
        }
    } else if (jsonObject) {
        var object = json; ///
        for(var name in object){
            var json2 = object[name], jsonString1 = isJSONString(json2);
            if (jsonString1) {
                var string1 = json2, value1 = replaceEnvironmentVariable(string1); ///
                if (value1 !== null) {
                    object[name] = value1;
                }
            } else {
                replaceEnvironmentVariables(json2);
            }
        }
    }
}
function isJSONArray(json) {
    var jsonArray = _instanceof(json, Array);
    return jsonArray;
}
function isJSONObject(json) {
    var jsonArray = isJSONArray(json), jsonPrimitive = isJSONPrimitive(json), jsonObject = !jsonArray && !jsonPrimitive;
    return jsonObject;
}
function isJSONString(json) {
    var jsonString = (typeof json === "undefined" ? "undefined" : _typeof(json)) === _constants.STRING;
    return jsonString;
}
function isJSONNumber(json) {
    var jsonNumber = (typeof json === "undefined" ? "undefined" : _typeof(json)) === _constants.NUMBER;
    return jsonNumber;
}
function isJSONBoolean(json) {
    var jsonBoolean = (typeof json === "undefined" ? "undefined" : _typeof(json)) === _constants.BOOLEAN;
    return jsonBoolean;
}
function isJSONPrimitive(json) {
    var jsonString = isJSONString(json), jsonNumber = isJSONNumber(json), jsonBoolean = isJSONBoolean(json), jsonPrimitive = jsonString || jsonNumber || jsonBoolean;
    return jsonPrimitive;
}
function isStringUpperCase(string) {
    var upperCaseString = string.toUpperCase(), stringUpperCase = string === upperCaseString;
    return stringUpperCase;
}
