// necessary/lib/utilities/request.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _constants = require("../constants");
var _methods = require("../methods");
var _http = require("../utilities/http");
var http = require("http"), https = require("https");
function createRequest(host, uri, query, method, headers, callback) {
    var port = (0, _http).portFromHost(host), secure = (0, _http).secureFromHost(host), hostname = (0, _http).hostnameFromHost(host), queryString = (0, _http).queryStringFromQuery(query), path = queryString === _constants.EMPTY_STRING ? uri : "".concat(uri, "?").concat(queryString), options = {
        hostname: hostname,
        headers: headers,
        method: method,
        port: port,
        path: path
    }, createRequest1 = secure ? https.request : http.request; ///
    var request = createRequest1(options, function(response) {
        var error = null;
        callback(error, response);
    });
    request.on(_constants.ERROR, function(error) {
        var response = null;
        callback(error, response);
    });
    return request;
}
function createGetRequest(host, uri, query, headers, callback) {
    if (callback === undefined) {
        callback = headers; ///
        headers = {};
    }
    var method = _methods.GET_METHOD, getRequest = createRequest(host, uri, query, method, headers, callback);
    return getRequest;
}
function createPostRequest(host, uri, query, headers, callback) {
    if (callback === undefined) {
        callback = headers; ///
        headers = {};
    }
    var method = _methods.POST_METHOD, postRequest = createRequest(host, uri, query, method, headers, callback);
    return postRequest;
}
var _default = {
    createRequest: createRequest,
    createGetRequest: createGetRequest,
    createPostRequest: createPostRequest
};
exports.default = _default;
