// necessary/lib/characters.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.EXCLAMATION_MARK_CHARACTER = exports.CARRIAGE_RETURN_CHARACTER = exports.FORWARD_SLASH_CHARACTER = exports.GREATER_THAN_CHARACTER = exports.LESS_THAN_CHARACTER = exports.AMPERSAND_CHARACTER = exports.BACKSPACE_CHARACTER = exports.NEW_LINE_CHARACTER = exports.BACKTICK_DELIMITER = exports.WILDCARD_CHARACTER = exports.CTRL_C_CHARACTER = exports.PERIOD_CHARACTER = exports.COLON_CHARACTER = exports.COMMA_CHARACTER = exports.SPACE_CHARACTER = exports.RIGHT_CHARACTER = exports.LEFT_CHARACTER = exports.DOWN_CHARACTER = exports.DASH_CHARACTER = exports.BAR_CHARACTER = exports.ETX_CHARACTER = exports.UP_CHARACTER = void 0;
var UP_CHARACTER = "\x1b[A";
exports.UP_CHARACTER = UP_CHARACTER;
var ETX_CHARACTER = "\x03";
exports.ETX_CHARACTER = ETX_CHARACTER;
var BAR_CHARACTER = "|";
exports.BAR_CHARACTER = BAR_CHARACTER;
var DASH_CHARACTER = "-";
exports.DASH_CHARACTER = DASH_CHARACTER;
var DOWN_CHARACTER = "\x1b[B";
exports.DOWN_CHARACTER = DOWN_CHARACTER;
var LEFT_CHARACTER = "\x1b[D";
exports.LEFT_CHARACTER = LEFT_CHARACTER;
var RIGHT_CHARACTER = "\x1b[C";
exports.RIGHT_CHARACTER = RIGHT_CHARACTER;
var SPACE_CHARACTER = " ";
exports.SPACE_CHARACTER = SPACE_CHARACTER;
var COMMA_CHARACTER = ",";
exports.COMMA_CHARACTER = COMMA_CHARACTER;
var COLON_CHARACTER = ":";
exports.COLON_CHARACTER = COLON_CHARACTER;
var PERIOD_CHARACTER = ".";
exports.PERIOD_CHARACTER = PERIOD_CHARACTER;
var CTRL_C_CHARACTER = "^C";
exports.CTRL_C_CHARACTER = CTRL_C_CHARACTER;
var WILDCARD_CHARACTER = "*";
exports.WILDCARD_CHARACTER = WILDCARD_CHARACTER;
var BACKTICK_DELIMITER = "`";
exports.BACKTICK_DELIMITER = BACKTICK_DELIMITER;
var NEW_LINE_CHARACTER = "\n";
exports.NEW_LINE_CHARACTER = NEW_LINE_CHARACTER;
var BACKSPACE_CHARACTER = String.fromCharCode(127);
exports.BACKSPACE_CHARACTER = BACKSPACE_CHARACTER;
var AMPERSAND_CHARACTER = "&";
exports.AMPERSAND_CHARACTER = AMPERSAND_CHARACTER;
var LESS_THAN_CHARACTER = "&lt;";
exports.LESS_THAN_CHARACTER = LESS_THAN_CHARACTER;
var GREATER_THAN_CHARACTER = "&gt;";
exports.GREATER_THAN_CHARACTER = GREATER_THAN_CHARACTER;
var FORWARD_SLASH_CHARACTER = "/";
exports.FORWARD_SLASH_CHARACTER = FORWARD_SLASH_CHARACTER;
var CARRIAGE_RETURN_CHARACTER = "\r";
exports.CARRIAGE_RETURN_CHARACTER = CARRIAGE_RETURN_CHARACTER;
var EXCLAMATION_MARK_CHARACTER = "!";
exports.EXCLAMATION_MARK_CHARACTER = EXCLAMATION_MARK_CHARACTER;
var _default = {
    UP_CHARACTER: UP_CHARACTER,
    ETX_CHARACTER: ETX_CHARACTER,
    BAR_CHARACTER: BAR_CHARACTER,
    DASH_CHARACTER: DASH_CHARACTER,
    DOWN_CHARACTER: DOWN_CHARACTER,
    LEFT_CHARACTER: LEFT_CHARACTER,
    RIGHT_CHARACTER: RIGHT_CHARACTER,
    SPACE_CHARACTER: SPACE_CHARACTER,
    COMMA_CHARACTER: COMMA_CHARACTER,
    COLON_CHARACTER: COLON_CHARACTER,
    PERIOD_CHARACTER: PERIOD_CHARACTER,
    CTRL_C_CHARACTER: CTRL_C_CHARACTER,
    WILDCARD_CHARACTER: WILDCARD_CHARACTER,
    BACKTICK_DELIMITER: BACKTICK_DELIMITER,
    NEW_LINE_CHARACTER: NEW_LINE_CHARACTER,
    AMPERSAND_CHARACTER: AMPERSAND_CHARACTER,
    BACKSPACE_CHARACTER: BACKSPACE_CHARACTER,
    LESS_THAN_CHARACTER: LESS_THAN_CHARACTER,
    GREATER_THAN_CHARACTER: GREATER_THAN_CHARACTER,
    FORWARD_SLASH_CHARACTER: FORWARD_SLASH_CHARACTER,
    CARRIAGE_RETURN_CHARACTER: CARRIAGE_RETURN_CHARACTER,
    EXCLAMATION_MARK_CHARACTER: EXCLAMATION_MARK_CHARACTER
};
exports.default = _default;
