// necessary/lib/contentTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE = exports.APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE = exports.TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE = exports.TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE = exports.APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE = exports.APPLICATION_OCTET_STREAM_CONTENT_TYPE = exports.APPLICATION_JSON_CONTENT_TYPE = exports.TEXT_PLAIN_CONTENT_TYPE = exports.TEXT_HTML_CONTENT_TYPE = void 0;
var TEXT_HTML_CONTENT_TYPE = "text/html";
exports.TEXT_HTML_CONTENT_TYPE = TEXT_HTML_CONTENT_TYPE;
var TEXT_PLAIN_CONTENT_TYPE = "text/plain";
exports.TEXT_PLAIN_CONTENT_TYPE = TEXT_PLAIN_CONTENT_TYPE;
var APPLICATION_JSON_CONTENT_TYPE = "application/json";
exports.APPLICATION_JSON_CONTENT_TYPE = APPLICATION_JSON_CONTENT_TYPE;
var APPLICATION_OCTET_STREAM_CONTENT_TYPE = "application/octet-stream";
exports.APPLICATION_OCTET_STREAM_CONTENT_TYPE = APPLICATION_OCTET_STREAM_CONTENT_TYPE;
var APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE = "application/x-www-form-urlencoded";
exports.APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE = APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE;
var TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE = "text/html; charset=utf-8";
exports.TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE = TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE;
var TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE = "text/plain; charset=utf-8";
exports.TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE = TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE;
var APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE = "application/json; charset=utf-8";
exports.APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE = APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE;
var APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE = "application/x-www-form-urlencoded; charset=utf-8";
exports.APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE = APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE;
var _default = {
    TEXT_HTML_CONTENT_TYPE: TEXT_HTML_CONTENT_TYPE,
    TEXT_PLAIN_CONTENT_TYPE: TEXT_PLAIN_CONTENT_TYPE,
    APPLICATION_JSON_CONTENT_TYPE: APPLICATION_JSON_CONTENT_TYPE,
    APPLICATION_OCTET_STREAM_CONTENT_TYPE: APPLICATION_OCTET_STREAM_CONTENT_TYPE,
    APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE: APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE,
    TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE: TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE,
    TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE: TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE,
    APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE: APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE,
    APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE: APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE
};
exports.default = _default;
