// necessary/lib/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.DEFAULT_LOG_FILE_BASE_NAME = exports.DEFAULT_LOG_DIRECTORY_PATH = exports.DEFAULT_RC_BASE_EXTENSION = exports.DEFAULT_INITIAL_ANSWER = exports.DEFAULT_LOG_LEVEL = exports.DEFAULT_ENCODING = exports.DEFAULT_ATTEMPTS = void 0;
var _levels = require("./levels");
var _encodings = require("./encodings");
var _constants = require("./constants");
var DEFAULT_ATTEMPTS = 3;
exports.DEFAULT_ATTEMPTS = DEFAULT_ATTEMPTS;
var DEFAULT_ENCODING = _encodings.UTF8_ENCODING;
exports.DEFAULT_ENCODING = DEFAULT_ENCODING;
var DEFAULT_LOG_LEVEL = _levels.WARNING_LEVEL;
exports.DEFAULT_LOG_LEVEL = DEFAULT_LOG_LEVEL;
var DEFAULT_INITIAL_ANSWER = _constants.EMPTY_STRING;
exports.DEFAULT_INITIAL_ANSWER = DEFAULT_INITIAL_ANSWER;
var DEFAULT_RC_BASE_EXTENSION = _constants.EMPTY_STRING;
exports.DEFAULT_RC_BASE_EXTENSION = DEFAULT_RC_BASE_EXTENSION;
var DEFAULT_LOG_DIRECTORY_PATH = null;
exports.DEFAULT_LOG_DIRECTORY_PATH = DEFAULT_LOG_DIRECTORY_PATH;
var DEFAULT_LOG_FILE_BASE_NAME = _constants.DEFAULT;
exports.DEFAULT_LOG_FILE_BASE_NAME = DEFAULT_LOG_FILE_BASE_NAME;
