// necessary/lib/headers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.ACCESS_CONTROL_REQUEST_METHOD_HEADER = exports.ACCESS_CONTROL_ALLOW_HEADERS_HEADER = exports.ACCESS_CONTROL_ALLOW_METHODS_HEADER = exports.ACCESS_CONTROL_ALLOW_ORIGIN_HEADER = exports.CONTENT_DISPOSITION_HEADER = exports.TRANSFER_ENCODING_HEADER = exports.CONTENT_LENGTH_HEADER = exports.CACHE_CONTROL_HEADER = exports.AUTHORIZATION_HEADER = exports.CONTENT_TYPE_HEADER = exports.USER_AGENT_HEADER = exports.LOCATION_HEADER = exports.ACCEPT_HEADER = exports.PRAGMA_HEADER = void 0;
var PRAGMA_HEADER = "pragma";
exports.PRAGMA_HEADER = PRAGMA_HEADER;
var ACCEPT_HEADER = "accept";
exports.ACCEPT_HEADER = ACCEPT_HEADER;
var LOCATION_HEADER = "location";
exports.LOCATION_HEADER = LOCATION_HEADER;
var USER_AGENT_HEADER = "user-agent";
exports.USER_AGENT_HEADER = USER_AGENT_HEADER;
var CONTENT_TYPE_HEADER = "content-type";
exports.CONTENT_TYPE_HEADER = CONTENT_TYPE_HEADER;
var AUTHORIZATION_HEADER = "authorization";
exports.AUTHORIZATION_HEADER = AUTHORIZATION_HEADER;
var CACHE_CONTROL_HEADER = "cache-control";
exports.CACHE_CONTROL_HEADER = CACHE_CONTROL_HEADER;
var CONTENT_LENGTH_HEADER = "content-length";
exports.CONTENT_LENGTH_HEADER = CONTENT_LENGTH_HEADER;
var TRANSFER_ENCODING_HEADER = "transfer-encoding";
exports.TRANSFER_ENCODING_HEADER = TRANSFER_ENCODING_HEADER;
var CONTENT_DISPOSITION_HEADER = "content-disposition";
exports.CONTENT_DISPOSITION_HEADER = CONTENT_DISPOSITION_HEADER;
var ACCESS_CONTROL_ALLOW_ORIGIN_HEADER = "access-control-allow-origin";
exports.ACCESS_CONTROL_ALLOW_ORIGIN_HEADER = ACCESS_CONTROL_ALLOW_ORIGIN_HEADER;
var ACCESS_CONTROL_ALLOW_METHODS_HEADER = "access-control-allow-methods";
exports.ACCESS_CONTROL_ALLOW_METHODS_HEADER = ACCESS_CONTROL_ALLOW_METHODS_HEADER;
var ACCESS_CONTROL_ALLOW_HEADERS_HEADER = "access-control-allow-headers";
exports.ACCESS_CONTROL_ALLOW_HEADERS_HEADER = ACCESS_CONTROL_ALLOW_HEADERS_HEADER;
var ACCESS_CONTROL_REQUEST_METHOD_HEADER = "access-control-request-method";
exports.ACCESS_CONTROL_REQUEST_METHOD_HEADER = ACCESS_CONTROL_REQUEST_METHOD_HEADER;
var _default = {
    PRAGMA_HEADER: PRAGMA_HEADER,
    ACCEPT_HEADER: ACCEPT_HEADER,
    LOCATION_HEADER: LOCATION_HEADER,
    USER_AGENT_HEADER: USER_AGENT_HEADER,
    CONTENT_TYPE_HEADER: CONTENT_TYPE_HEADER,
    AUTHORIZATION_HEADER: AUTHORIZATION_HEADER,
    CACHE_CONTROL_HEADER: CACHE_CONTROL_HEADER,
    CONTENT_LENGTH_HEADER: CONTENT_LENGTH_HEADER,
    TRANSFER_ENCODING_HEADER: TRANSFER_ENCODING_HEADER,
    CONTENT_DISPOSITION_HEADER: CONTENT_DISPOSITION_HEADER,
    ACCESS_CONTROL_ALLOW_ORIGIN_HEADER: ACCESS_CONTROL_ALLOW_ORIGIN_HEADER,
    ACCESS_CONTROL_ALLOW_METHODS_HEADER: ACCESS_CONTROL_ALLOW_METHODS_HEADER,
    ACCESS_CONTROL_ALLOW_HEADERS_HEADER: ACCESS_CONTROL_ALLOW_HEADERS_HEADER,
    ACCESS_CONTROL_REQUEST_METHOD_HEADER: ACCESS_CONTROL_REQUEST_METHOD_HEADER
};
exports.default = _default;
