// necessary/lib/statusCodes.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.INTERNAL_SERVER_ERROR_500_STATUS_CODE = exports.SERVICE_UNAVAILABLE_503_STATUS_CODE = exports.TOO_MANY_REQUESTS_429_STATUS_CODE = exports.REQUEST_TIMEOUT_408_STATUS_CODE = exports.UNAUTHORIZED_401_STATUS_CODE = exports.BAD_REQUEST_400_STATUS_CODE = exports.BAD_GATEWAY_502_STATUS_CODE = exports.NO_CONTENT_204_STATUS_CODE = exports.NOT_FOUND_404_STATUS_CODE = exports.FORBIDDEN_403_STATUS_CODE = exports.SEE_OTHER_303_STATUS_CODE = exports.CREATED_201_STATUS_CODE = exports.FOUND_302_STATUS_CODE = exports.OK_200_STATUS_CODE = exports.ZERO_0_STATUS_CODE = void 0;
var ZERO_0_STATUS_CODE = 0;
exports.ZERO_0_STATUS_CODE = ZERO_0_STATUS_CODE;
var OK_200_STATUS_CODE = 200;
exports.OK_200_STATUS_CODE = OK_200_STATUS_CODE;
var FOUND_302_STATUS_CODE = 302;
exports.FOUND_302_STATUS_CODE = FOUND_302_STATUS_CODE;
var CREATED_201_STATUS_CODE = 201;
exports.CREATED_201_STATUS_CODE = CREATED_201_STATUS_CODE;
var SEE_OTHER_303_STATUS_CODE = 303;
exports.SEE_OTHER_303_STATUS_CODE = SEE_OTHER_303_STATUS_CODE;
var FORBIDDEN_403_STATUS_CODE = 403;
exports.FORBIDDEN_403_STATUS_CODE = FORBIDDEN_403_STATUS_CODE;
var NOT_FOUND_404_STATUS_CODE = 404;
exports.NOT_FOUND_404_STATUS_CODE = NOT_FOUND_404_STATUS_CODE;
var NO_CONTENT_204_STATUS_CODE = 204;
exports.NO_CONTENT_204_STATUS_CODE = NO_CONTENT_204_STATUS_CODE;
var BAD_GATEWAY_502_STATUS_CODE = 502;
exports.BAD_GATEWAY_502_STATUS_CODE = BAD_GATEWAY_502_STATUS_CODE;
var BAD_REQUEST_400_STATUS_CODE = 400;
exports.BAD_REQUEST_400_STATUS_CODE = BAD_REQUEST_400_STATUS_CODE;
var UNAUTHORIZED_401_STATUS_CODE = 401;
exports.UNAUTHORIZED_401_STATUS_CODE = UNAUTHORIZED_401_STATUS_CODE;
var REQUEST_TIMEOUT_408_STATUS_CODE = 408;
exports.REQUEST_TIMEOUT_408_STATUS_CODE = REQUEST_TIMEOUT_408_STATUS_CODE;
var TOO_MANY_REQUESTS_429_STATUS_CODE = 429;
exports.TOO_MANY_REQUESTS_429_STATUS_CODE = TOO_MANY_REQUESTS_429_STATUS_CODE;
var SERVICE_UNAVAILABLE_503_STATUS_CODE = 503;
exports.SERVICE_UNAVAILABLE_503_STATUS_CODE = SERVICE_UNAVAILABLE_503_STATUS_CODE;
var INTERNAL_SERVER_ERROR_500_STATUS_CODE = 500;
exports.INTERNAL_SERVER_ERROR_500_STATUS_CODE = INTERNAL_SERVER_ERROR_500_STATUS_CODE;
var _default = {
    ZERO_0_STATUS_CODE: ZERO_0_STATUS_CODE,
    OK_200_STATUS_CODE: OK_200_STATUS_CODE,
    FOUND_302_STATUS_CODE: FOUND_302_STATUS_CODE,
    CREATED_201_STATUS_CODE: CREATED_201_STATUS_CODE,
    SEE_OTHER_303_STATUS_CODE: SEE_OTHER_303_STATUS_CODE,
    FORBIDDEN_403_STATUS_CODE: FORBIDDEN_403_STATUS_CODE,
    NOT_FOUND_404_STATUS_CODE: NOT_FOUND_404_STATUS_CODE,
    NO_CONTENT_204_STATUS_CODE: NO_CONTENT_204_STATUS_CODE,
    BAD_GATEWAY_502_STATUS_CODE: BAD_GATEWAY_502_STATUS_CODE,
    BAD_REQUEST_400_STATUS_CODE: BAD_REQUEST_400_STATUS_CODE,
    UNAUTHORIZED_401_STATUS_CODE: UNAUTHORIZED_401_STATUS_CODE,
    REQUEST_TIMEOUT_408_STATUS_CODE: REQUEST_TIMEOUT_408_STATUS_CODE,
    TOO_MANY_REQUESTS_429_STATUS_CODE: TOO_MANY_REQUESTS_429_STATUS_CODE,
    SERVICE_UNAVAILABLE_503_STATUS_CODE: SERVICE_UNAVAILABLE_503_STATUS_CODE,
    INTERNAL_SERVER_ERROR_500_STATUS_CODE: INTERNAL_SERVER_ERROR_500_STATUS_CODE
};
exports.default = _default;
