// necessary/lib/statusMessages.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE = exports.SERVICE_UNAVAILABLE_503_STATUS_MESSAGE = exports.TOO_MANY_REQUESTS_429_STATUS_MESSAGE = exports.REQUEST_TIMEOUT_408_STATUS_MESSAGE = exports.UNAUTHORIZED_401_STATUS_MESSAGE = exports.BAD_REQUEST_400_STATUS_MESSAGE = exports.BAD_GATEWAY_502_STATUS_MESSAGE = exports.NO_CONTENT_204_STATUS_MESSAGE = exports.NOT_FOUND_404_STATUS_MESSAGE = exports.FORBIDDEN_403_STATUS_MESSAGE = exports.SEE_OTHER_303_STATUS_MESSAGE = exports.CREATED_201_STATUS_MESSAGE = exports.FOUND_302_STATUS_MESSAGE = exports.OK_200_STATUS_MESSAGE = exports.ZERO_0_STATUS_MESSAGE = void 0;
var ZERO_0_STATUS_MESSAGE = "";
exports.ZERO_0_STATUS_MESSAGE = ZERO_0_STATUS_MESSAGE;
var OK_200_STATUS_MESSAGE = "OK";
exports.OK_200_STATUS_MESSAGE = OK_200_STATUS_MESSAGE;
var FOUND_302_STATUS_MESSAGE = "Found";
exports.FOUND_302_STATUS_MESSAGE = FOUND_302_STATUS_MESSAGE;
var CREATED_201_STATUS_MESSAGE = "Created";
exports.CREATED_201_STATUS_MESSAGE = CREATED_201_STATUS_MESSAGE;
var SEE_OTHER_303_STATUS_MESSAGE = "See other";
exports.SEE_OTHER_303_STATUS_MESSAGE = SEE_OTHER_303_STATUS_MESSAGE;
var FORBIDDEN_403_STATUS_MESSAGE = "Forbidden";
exports.FORBIDDEN_403_STATUS_MESSAGE = FORBIDDEN_403_STATUS_MESSAGE;
var NOT_FOUND_404_STATUS_MESSAGE = "Not found";
exports.NOT_FOUND_404_STATUS_MESSAGE = NOT_FOUND_404_STATUS_MESSAGE;
var NO_CONTENT_204_STATUS_MESSAGE = "No content";
exports.NO_CONTENT_204_STATUS_MESSAGE = NO_CONTENT_204_STATUS_MESSAGE;
var BAD_GATEWAY_502_STATUS_MESSAGE = "Bad gateway";
exports.BAD_GATEWAY_502_STATUS_MESSAGE = BAD_GATEWAY_502_STATUS_MESSAGE;
var BAD_REQUEST_400_STATUS_MESSAGE = "Bad request";
exports.BAD_REQUEST_400_STATUS_MESSAGE = BAD_REQUEST_400_STATUS_MESSAGE;
var UNAUTHORIZED_401_STATUS_MESSAGE = "Unauthorized";
exports.UNAUTHORIZED_401_STATUS_MESSAGE = UNAUTHORIZED_401_STATUS_MESSAGE;
var REQUEST_TIMEOUT_408_STATUS_MESSAGE = "Request timeout";
exports.REQUEST_TIMEOUT_408_STATUS_MESSAGE = REQUEST_TIMEOUT_408_STATUS_MESSAGE;
var TOO_MANY_REQUESTS_429_STATUS_MESSAGE = "Too many requests";
exports.TOO_MANY_REQUESTS_429_STATUS_MESSAGE = TOO_MANY_REQUESTS_429_STATUS_MESSAGE;
var SERVICE_UNAVAILABLE_503_STATUS_MESSAGE = "Service unavailable";
exports.SERVICE_UNAVAILABLE_503_STATUS_MESSAGE = SERVICE_UNAVAILABLE_503_STATUS_MESSAGE;
var INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE = "Internal server error";
exports.INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE = INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE;
var _default = {
    ZERO_0_STATUS_MESSAGE: ZERO_0_STATUS_MESSAGE,
    OK_200_STATUS_MESSAGE: OK_200_STATUS_MESSAGE,
    FOUND_302_STATUS_MESSAGE: FOUND_302_STATUS_MESSAGE,
    CREATED_201_STATUS_MESSAGE: CREATED_201_STATUS_MESSAGE,
    SEE_OTHER_303_STATUS_MESSAGE: SEE_OTHER_303_STATUS_MESSAGE,
    FORBIDDEN_403_STATUS_MESSAGE: FORBIDDEN_403_STATUS_MESSAGE,
    NOT_FOUND_404_STATUS_MESSAGE: NOT_FOUND_404_STATUS_MESSAGE,
    NO_CONTENT_204_STATUS_MESSAGE: NO_CONTENT_204_STATUS_MESSAGE,
    BAD_GATEWAY_502_STATUS_MESSAGE: BAD_GATEWAY_502_STATUS_MESSAGE,
    BAD_REQUEST_400_STATUS_MESSAGE: BAD_REQUEST_400_STATUS_MESSAGE,
    UNAUTHORIZED_401_STATUS_MESSAGE: UNAUTHORIZED_401_STATUS_MESSAGE,
    REQUEST_TIMEOUT_408_STATUS_MESSAGE: REQUEST_TIMEOUT_408_STATUS_MESSAGE,
    TOO_MANY_REQUESTS_429_STATUS_MESSAGE: TOO_MANY_REQUESTS_429_STATUS_MESSAGE,
    SERVICE_UNAVAILABLE_503_STATUS_MESSAGE: SERVICE_UNAVAILABLE_503_STATUS_MESSAGE,
    INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE: INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE
};
exports.default = _default;
