// necessary/src/characters.js
"use strict";

export const UP_CHARACTER = "[A";
export const ETX_CHARACTER = "\u0003";
export const BAR_CHARACTER = "|";
export const DASH_CHARACTER = "-";
export const DOWN_CHARACTER = "[B";
export const LEFT_CHARACTER = "[D";
export const RIGHT_CHARACTER = "[C";
export const SPACE_CHARACTER = " ";
export const COMMA_CHARACTER = ",";
export const COLON_CHARACTER = ":";
export const PERIOD_CHARACTER = ".";
export const CTRL_C_CHARACTER = "^C";
export const WILDCARD_CHARACTER = "*";
export const BACKTICK_DELIMITER = "`";
export const NEW_LINE_CHARACTER = "\n";
export const BACKSPACE_CHARACTER = String.fromCharCode(127);
export const AMPERSAND_CHARACTER = "&";
export const LESS_THAN_CHARACTER = "&lt;";
export const GREATER_THAN_CHARACTER = "&gt;";
export const FORWARD_SLASH_CHARACTER = "/";
export const CARRIAGE_RETURN_CHARACTER = "\r";
export const EXCLAMATION_MARK_CHARACTER = "!";

export default {
  UP_CHARACTER,
  ETX_CHARACTER,
  BAR_CHARACTER,
  DASH_CHARACTER,
  DOWN_CHARACTER,
  LEFT_CHARACTER,
  RIGHT_CHARACTER,
  SPACE_CHARACTER,
  COMMA_CHARACTER,
  COLON_CHARACTER,
  PERIOD_CHARACTER,
  CTRL_C_CHARACTER,
  WILDCARD_CHARACTER,
  BACKTICK_DELIMITER,
  NEW_LINE_CHARACTER,
  AMPERSAND_CHARACTER,
  BACKSPACE_CHARACTER,
  LESS_THAN_CHARACTER,
  GREATER_THAN_CHARACTER,
  FORWARD_SLASH_CHARACTER,
  CARRIAGE_RETURN_CHARACTER,
  EXCLAMATION_MARK_CHARACTER
};
