// necessary/src/contentTypes.js
"use strict";

export const TEXT_HTML_CONTENT_TYPE = "text/html";
export const TEXT_PLAIN_CONTENT_TYPE = "text/plain";
export const APPLICATION_JSON_CONTENT_TYPE = "application/json";
export const APPLICATION_OCTET_STREAM_CONTENT_TYPE = "application/octet-stream";
export const APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE = "application/x-www-form-urlencoded";
export const TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE = "text/html; charset=utf-8";
export const TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE = "text/plain; charset=utf-8";
export const APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE = "application/json; charset=utf-8";
export const APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE = "application/x-www-form-urlencoded; charset=utf-8";

export default {
  TEXT_HTML_CONTENT_TYPE,
  TEXT_PLAIN_CONTENT_TYPE,
  APPLICATION_JSON_CONTENT_TYPE,
  APPLICATION_OCTET_STREAM_CONTENT_TYPE,
  APPLICATION_JSON_CHARSET_UTF_8_CONTENT_TYPE,
  TEXT_HTML_CHARSET_UTF_8_CONTENT_TYPE,
  TEXT_PLAIN_CHARSET_UTF_8_CONTENT_TYPE,
  APPLICATION_X_WWW_FORM_ENCODED_CONTENT_TYPE,
  APPLICATION_X_WWW_FORM_ENCODED_CHARSET_UTF_8_CONTENT_TYPE
};
