// necessary/src/headers.js
"use strict";

export const PRAGMA_HEADER = "pragma";
export const ACCEPT_HEADER = "accept";
export const LOCATION_HEADER = "location";
export const USER_AGENT_HEADER = "user-agent";
export const CONTENT_TYPE_HEADER = "content-type";
export const AUTHORIZATION_HEADER = "authorization";
export const CACHE_CONTROL_HEADER = "cache-control";
export const CONTENT_LENGTH_HEADER = "content-length";
export const TRANSFER_ENCODING_HEADER = "transfer-encoding";
export const CONTENT_DISPOSITION_HEADER = "content-disposition";
export const ACCESS_CONTROL_ALLOW_ORIGIN_HEADER = "access-control-allow-origin";
export const ACCESS_CONTROL_ALLOW_METHODS_HEADER = "access-control-allow-methods";
export const ACCESS_CONTROL_ALLOW_HEADERS_HEADER = "access-control-allow-headers";
export const ACCESS_CONTROL_REQUEST_METHOD_HEADER = "access-control-request-method";

export default {
  PRAGMA_HEADER,
  ACCEPT_HEADER,
  LOCATION_HEADER,
  USER_AGENT_HEADER,
  CONTENT_TYPE_HEADER,
  AUTHORIZATION_HEADER,
  CACHE_CONTROL_HEADER,
  CONTENT_LENGTH_HEADER,
  TRANSFER_ENCODING_HEADER,
  CONTENT_DISPOSITION_HEADER,
  ACCESS_CONTROL_ALLOW_ORIGIN_HEADER,
  ACCESS_CONTROL_ALLOW_METHODS_HEADER,
  ACCESS_CONTROL_ALLOW_HEADERS_HEADER,
  ACCESS_CONTROL_REQUEST_METHOD_HEADER
};
