// necessary/src/statusCodes.js
"use strict";

export const ZERO_0_STATUS_CODE = 0;
export const OK_200_STATUS_CODE = 200;
export const FOUND_302_STATUS_CODE = 302;
export const CREATED_201_STATUS_CODE = 201;
export const SEE_OTHER_303_STATUS_CODE = 303;
export const FORBIDDEN_403_STATUS_CODE = 403;
export const NOT_FOUND_404_STATUS_CODE = 404;
export const NO_CONTENT_204_STATUS_CODE = 204;
export const BAD_GATEWAY_502_STATUS_CODE = 502;
export const BAD_REQUEST_400_STATUS_CODE = 400;
export const UNAUTHORIZED_401_STATUS_CODE = 401;
export const REQUEST_TIMEOUT_408_STATUS_CODE = 408;
export const TOO_MANY_REQUESTS_429_STATUS_CODE = 429;
export const SERVICE_UNAVAILABLE_503_STATUS_CODE = 503;
export const INTERNAL_SERVER_ERROR_500_STATUS_CODE = 500;

export default {
  ZERO_0_STATUS_CODE,
  OK_200_STATUS_CODE,
  FOUND_302_STATUS_CODE,
  CREATED_201_STATUS_CODE,
  SEE_OTHER_303_STATUS_CODE,
  FORBIDDEN_403_STATUS_CODE,
  NOT_FOUND_404_STATUS_CODE,
  NO_CONTENT_204_STATUS_CODE,
  BAD_GATEWAY_502_STATUS_CODE,
  BAD_REQUEST_400_STATUS_CODE,
  UNAUTHORIZED_401_STATUS_CODE,
  REQUEST_TIMEOUT_408_STATUS_CODE,
  TOO_MANY_REQUESTS_429_STATUS_CODE,
  SERVICE_UNAVAILABLE_503_STATUS_CODE,
  INTERNAL_SERVER_ERROR_500_STATUS_CODE
};
