// necessary/src/statusMessages.js
"use strict";

export const ZERO_0_STATUS_MESSAGE = "";
export const OK_200_STATUS_MESSAGE = "OK";
export const FOUND_302_STATUS_MESSAGE = "Found";
export const CREATED_201_STATUS_MESSAGE = "Created";
export const SEE_OTHER_303_STATUS_MESSAGE = "See other";
export const FORBIDDEN_403_STATUS_MESSAGE = "Forbidden";
export const NOT_FOUND_404_STATUS_MESSAGE = "Not found";
export const NO_CONTENT_204_STATUS_MESSAGE = "No content";
export const BAD_GATEWAY_502_STATUS_MESSAGE = "Bad gateway";
export const BAD_REQUEST_400_STATUS_MESSAGE = "Bad request";
export const UNAUTHORIZED_401_STATUS_MESSAGE = "Unauthorized";
export const REQUEST_TIMEOUT_408_STATUS_MESSAGE = "Request timeout";
export const TOO_MANY_REQUESTS_429_STATUS_MESSAGE = "Too many requests";
export const SERVICE_UNAVAILABLE_503_STATUS_MESSAGE = "Service unavailable";
export const INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE = "Internal server error";

export default {
  ZERO_0_STATUS_MESSAGE,
  OK_200_STATUS_MESSAGE,
  FOUND_302_STATUS_MESSAGE,
  CREATED_201_STATUS_MESSAGE,
  SEE_OTHER_303_STATUS_MESSAGE,
  FORBIDDEN_403_STATUS_MESSAGE,
  NOT_FOUND_404_STATUS_MESSAGE,
  NO_CONTENT_204_STATUS_MESSAGE,
  BAD_GATEWAY_502_STATUS_MESSAGE,
  BAD_REQUEST_400_STATUS_MESSAGE,
  UNAUTHORIZED_401_STATUS_MESSAGE,
  REQUEST_TIMEOUT_408_STATUS_MESSAGE,
  TOO_MANY_REQUESTS_429_STATUS_MESSAGE,
  SERVICE_UNAVAILABLE_503_STATUS_MESSAGE,
  INTERNAL_SERVER_ERROR_500_STATUS_MESSAGE
};
